src/appleseed/renderer/kernel/rendering/final/adaptivepixelrenderer.cpp

```diff
@@ -187,16 +187,16 @@ namespace
 
             while (true)
             {
-                trackers[0].reset_variation();
-                trackers[1].reset_variation();
-                trackers[2].reset_variation();
-
                 // Don't exceed 'max' samples in total.
                 assert(trackers[0].get_size() <= m_params.m_max_samples);
                 const size_t remaining_samples = m_params.m_max_samples - trackers[0].get_size();
                 if (remaining_samples == 0)
                     break;
 
+                trackers[0].reset_variation();
```

Comment:
Yes, that looks correct actually.

I would go even further: I would move the other termination checks just after the check on the remaining number of samples:

    // Stop if the variation criterion is met.
    if (trackers[0].get_variation() <= m_params.m_max_variation &&
        trackers[1].get_variation() <= m_params.m_max_variation &&
        trackers[2].get_variation() <= m_params.m_max_variation)
        break;

Reply:
If you do so, sampling will stop on the first step. Just have to check the size at the same time.

Reply:
Oh you're right, sorry. Maybe it's better to leave the test where it was then. Sorry for the troubles!

Reply:
Ok no problem :)